### PR TITLE
INS-3655: recover panics in contracts

### PR DESCRIPTION
**Pull request by bigbes**
* also make panics in contract execution as user error (store it in DB
  as part of response)

### CODE REVIEW

File: logicrunner/preprocessor/templates/wrapper.go.tpl [hunk 1]
```diff
@@ -94,40 +96,69 @@ func INSMETHOD_GetPrototype(object []byte, data []byte) ([]byte, []byte, error)
...
+		)
+	}
+
+	needRecover := true
+	defer func() {
+		if !needRecover {
+			return
+		}
+		if r := recover(); r != nil {
+			recoveredError := errors.Wrap(errors.Errorf("%v", r), "Failed to execute method (panic)")
```
Comment (ruz) [inline logicrunner/preprocessor/templates/wrapper.go.tpl]: can you move this repeated code into a function?

Comment (bigbes) [reply logicrunner/preprocessor/templates/wrapper.go.tpl]: i cannot, it contains templated part :(

File: logicrunner/preprocessor/main_test.go [hunk 2]
```diff
@@ -597,47 +597,6 @@ func ( A ) Get() error {
 	s.NotContains(bufWrapper.String(), `"some/test/import/path"`)
 }
 
-func (s *PreprocessorSuite) TestImportsFromContractUseForReturnValue() {
```
Comment (ruz) [inline logicrunner/preprocessor/main_test.go]: I don't understand why you deleted this test. Can you explain?

Comment (bigbes) [reply logicrunner/preprocessor/main_test.go]: We're importing packages needed for declaration of contract return values, now. That test tests that we won't import that packages.

Comment (ruz) [reply logicrunner/preprocessor/main_test.go]: Nope, that test was checking that we deal correctly with return values of type declared in an external package. This is what it was testing. If you changed the way it works then you should change the test. Don't delete it.

File: logicrunner/preprocessor/templates/wrapper.go.tpl [hunk 3]
```diff
@@ -94,40 +96,69 @@ func INSMETHOD_GetPrototype(object []byte, data []byte) ([]byte, []byte, error)
...
+			recoveredError = ph.MakeErrorSerializable(recoveredError)
+
+			if PanicIsLogicalError {
+				ret{{ $method.LastErrorInRes }} = recoveredError
+
+				newState = object
+				err = serializeResults()
```
Comment (ruz) [inline logicrunner/preprocessor/templates/wrapper.go.tpl]: this part doesn't look correct, it should be:
```
err = serializeResults()
if err != nil {
    newState = object
}
```

File: logicrunner/builtin/builtin.go [hunk 4]
```diff
@@ -106,13 +112,23 @@ func (b *BuiltIn) CallConstructor(
...
+	data []byte,
+	method string,
+	args insolar.Arguments,
+) (
+	[]byte, insolar.Arguments, error,
+) {
 	executeStart := time.Now()
 	ctx = insmetrics.InsertTag(ctx, metrics.TagContractPrototype, b.PrototypeRefRegistry[codeRef])
-	ctx = insmetrics.InsertTag(ctx, metrics.TagContractMethodName, method)
+	ctx = insmetrics.InsertTag(ctx, metrics.TagContractMethodName, "Constructor")
```
Comment (ruz) [inline logicrunner/builtin/builtin.go]: You may have multiple constructors with different names. I don't mind a prefix or something else added here.  Please split out these changes to metrics from this PR. They have nothing to do with recovering panics in contracts.

Comment (bigbes) [reply logicrunner/builtin/builtin.go]: fixed

File: logicrunner/preprocessor/templates/wrapper.go.tpl (hunk 3 above)
Comment (bigbes) [reply logicrunner/preprocessor/templates/wrapper.go.tpl]: fixed

File: logicrunner/preprocessor/main_test.go (hunk 2 above)
Comment (bigbes) [reply logicrunner/preprocessor/main_test.go]: fixed

**Pull request merged**